notebooks/ir_scattered_light_manual_corrections/Correcting_for_Scattered_Light_in_IR_Exposures_by_Manually_Subtracting_Bad_Reads.ipynb

```diff
@@ -0,0 +1,963 @@
+   "metadata": {},
+   "source": [
+    "#### STEP 2: Correct the IMA product"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Here we walk through the process of subtracting the reads affected by Earth-limb scattered light from the cumulative science extension of the IMA. While an FLT product is also created in step 1, we do not use this file due to the poor-quality ramp fit in the presence of TVB.\n",
+    "\n",
+    "Overall, the steps are as follows:\n",
+    "\n",
+    "1. Grab the science data as a 3D array and the integration time as a 1D array from the science and time extensions of the IMA file and the associated dark reference file. \n",
+    "2. Convert science data from electrons/second to electrons.\n",
+    "3. Compute the difference between reads (the cumulative difference, in this case) to isolate counts and integration time from bad reads.\n",
+    "4. For each bad read, subtract the counts (and time) from ONLY that read from the total data."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ima = fits.open(raw_filepath.replace('raw', 'ima'))\n",
+    "\n",
+    "#grab the science data as a 3D array and the integration time as a 1D array\n",
+    "cube, integ_time = diff.read_wfc3(raw_filepath.replace('raw','ima'))\n",
+    "\n",
+    "\n",
+    "#do the same for the corresponding dark image. The dark reference file used for the observation can be found in the primary header\n",
+    "dark_file = ima[0].header['DARKFILE'].replace('iref$', os.getenv('iref')+'/')\n",
+    "dark_cube, dark_time = diff.read_wfc3(dark_file)\n",
+    "\n",
+    "\n",
+    "#convert from count rate to counts\n",
+    "cube_counts = np.zeros(np.shape(cube))\n",
+    "dark_counts = np.zeros(np.shape(dark_cube))\n",
+    "for img in range(cube.shape[2]):\n",
+    "    cube_counts[:,:,img] = cube[:,:,img]*integ_time[img]\n",
+    "    dark_counts[:,:,img ] = dark_cube[:,:,img]/dark_time[img]*integ_time[img]\n",
+    "    \n",
+    "    \n",
+    "#Compute the difference between reads (the cumulative difference, in this case) to isolate counts (and integration time) from bad reads\n",
+    "difference = np.diff(cube_counts, axis = 2)\n",
+    "dark_diff = np.diff(dark_counts, axis = 2)\n",
+    "dt = np.diff(integ_time)\n",
+    "\n",
+    "#For each bad read, subtract the counts (or time) from ONLY that read from the total data\n",
+    "final_dark = dark_counts[:,:,-1]*1  \n",
+    "final_sci = cube_counts[:,:,-1]*1\n",
+    "final_time = np.ones((1024, 1024))*integ_time[-1]\n",
+    "if (len(bad_reads) > 0):\n",
+    "    for read in bad_reads:\n",
+    "        index = cube_counts.shape[2]-read-1 #because the reads are stored in reverse order\n",
+    "        final_sci -= difference[:,:,index]\n",
+    "        final_dark -= dark_diff[:,:,index]\n",
+    "        final_time -= dt[index]\n",
+    "        "
```

Comment:
Sorry, I am a bit confused by this comment! I don't see the nan-filled array

Reply:
No problem! So the first element of `dark_time` and `integ_time` is 0. That means when you create `dark_counts`:
```
    dark_counts[:,:,img] = dark_cube[:,:,img]/dark_time[img]*integ_time[img]
```
you're dividing `dark_cube[:,:,0]` (a 1024x1024 array) by `dark_time[0]` (which is 0). That makes `dark_counts[:,:,0]` a NAN filled array since you divided by 0. Then, you create `dark_diff`:
```
dark_diff = np.diff(dark_counts, axis=2)
```
for which `dark_counts[:,:,0]` (the NAN filled array) is subtracted from `dark_counts[:,:,1]`. This means `dark_diff[:,:,0]` is also a NAN filled array. Next, you create `final_dark` by looping through `bad_reads=[11,12,13,14,15]`:
```
        index = cube_counts.shape[2]-read-1
        final_dark -= dark_diff[:,:,index]
```
`cube_counts.shape[2] = 16` so on the last iteration, `index = 16 - 15 - 1` which is 0. Since `dark_diff[:,:,0]` is subtracted from `final_dark`, it also becomes a NAN filled array.

If you already corrected the first part, then you should be all set, but this reply is assumes you're curious how the NANs form and propagates throughout your code. It also propagates to `signal` and `final_err` (which is mentioned in a later comment). If you're still confused, let me know and we can link up/chat 👍🏾

Reply:
Ohhhh I see! the reason I wasnt getting the NAN was becuase I changed the calculation (becuase dark_time and inted_time where the same)